compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java

```diff
@@ -2308,112 +2338,214 @@ private void analyzeArgIsolatedness(BLangInvocation invocationExpr, List<BLangEx
 
             if (reqArgCount < paramsCount) {
                 // Part of the non-rest params are provided via the vararg.
-                BTupleType tupleType = varArgType.tag == TypeTags.ARRAY ?
-                    getRepresentativeTupleTypeForRemainingArgs(paramsCount, reqArgCount, (BArrayType) varArgType) :
-                    (BTupleType) varArgType;
-
-                List<BType> memberTypes = tupleType.getTupleTypes();
+                if (varArgType.tag == TypeTags.ARRAY || varArgType.tag == TypeTags.TUPLE) {
+                    analyzeListConstructorRestArgs(invocationExpr, symbol, expectsIsolation, params, 
+                            paramsCount, reqArgCount, varArg, varArgType, varArgPos);
+                    this.listConstructorExprs = null;
+                    return;
+                }
+                analyzeMappingConstructorRestArgs(invocationExpr, symbol, expectsIsolation, params, 
+                        paramsCount, reqArgCount, varArg, varArgType, varArgPos);
+                this.mappingConstructorExprs = null;
+                return;
+            }
+        }
 
-                BLangExpression varArgExpr = varArg.expr;
-                boolean listConstrVarArg =  varArgExpr.getKind() == NodeKind.LIST_CONSTRUCTOR_EXPR;
-                BLangListConstructorExpr listConstructorExpr = listConstrVarArg ?
-                        (BLangListConstructorExpr) varArgExpr : null;
+        if (!Symbols.isFlagOn(symbol.restParam.flags, Flags.ISOLATED_PARAM)) {
+            for (BLangExpression restArg : restArgs) {
+                analyzeNode(restArg, env);
+            }
+            return;
+        }
 
-                if (!listConstrVarArg) {
-                    analyzeNode(varArg, env);
-                }
+        // Args for rest param provided as both individual args and vararg.
+        analyzeRestArgsForRestParam(invocationExpr, restArgs, symbol, expectsIsolation);
+    }
 
-                int tupleIndex = 0;
-                for (int i = reqArgCount; i < paramsCount; i++) {
-                    if (!Symbols.isFlagOn(params.get(i).flags, Flags.ISOLATED_PARAM)) {
-                        if (listConstrVarArg) {
-                            analyzeNode(listConstructorExpr.exprs.get(tupleIndex), env);
+    private void analyzeMappingConstructorRestArgs(BLangInvocation invocationExpr, BInvokableSymbol symbol,
+                                                boolean expectsIsolation, List<BVarSymbol> params, int paramsCount,
+                                                int reqArgCount, BLangRestArgsExpression varArg, BType varArgType,
+                                                Location varArgPos) {
+        BLangExpression varArgExpr = varArg.expr;
+        if (varArgType.tag != TypeTags.RECORD) {
+            return;
```

Comment:
There seem to be multiple codecov warnings?

Reply:
This check is to be on the safe side for all cases. Anyway, compiler won't allow mapping constructor values other than record type to pass as function arguments.

Reply:
We can go ahead and cast then.